### ISSUE TITLE
Improve performance when querying tables with many transactions in Azure 

### ISSUE THREAD

**Issue by kyrre** [delta-io/delta-kernel-rs#534]
### Please describe why this is necessary.

Querying tables in Azure Data Lake Storage that have a lot of transactions takes forever. 

This seems to happen because the `object_store` crate is not able to simply list the blobs that were created after the latest checkpoint. This appears to be a limitation in the Azure API.

This does _not_ happen when using Apache Spark (Databricks). In this case the operation finishes almost instantly, but it's not clear what they do differently. 

### Describe the functionality you are proposing.

I can think of two options that might work:

1. There might another Azure API we can use - maybe filter blobs on last modified time?
2. If we first retrieve the latest checkpoint then maybe it's possibly to then use a prefix filter, e.g., 0000000000000717*

### Additional context

This will happen whenever the table is a streaming destination.

**Comment by zachschuermann** [comment]
Hi @kyrre! Thanks for opening this issue! One thing that could help us move forward on this front is to have a reproducible example (e.g. a table in Azure that we can observe takes 5s to read with delta-spark and 30s to read with kernel). Do you have a chance to look into that?

Also, yea, I recall there being some limitations with ADLS listing would be useful to document them here (and any of the limitations exposed via `object_store`)

**Comment by scovich** [comment]
Yes, IIRC ADLS listing API is very limited compared to S3 or GCS. In particular, you can't specify a lower bound on the listing, so it just lists the whole directory every time. Given that this doesn't seem to be a problem in spark, we should check what the ADLS hadoop client is doing. If it has some clever workaround, we should file an issue upstream for `object_store` to incorporate that same trick.

Unfortunately, I don't know that there's a lot we could do from the kernel side, if `object_store` doesn't make this efficient -- kernel is not generally in the business of solving cloud store API issues.

**Comment by sherlockbeard** [comment]
https://github.com/delta-io/delta/issues/1568 jvm version of this ticket

**Comment by sebbegg** [comment]
fwiw, azure supports the `startFrom` query parameter since some time in its listing API:
https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs?tabs=microsoft-entra-id#uri-parameters

**Comment by scovich** [comment]
@sebbegg -- Last time I looked at the docs for Azure `startFrom`, I came away with the impression that it was "well-nested" (= not helpful). But re-reading now maybe I mixed up its references to the `prefix` param (which _is_ well-nested by definition)?

Kernel just needs the ability to list from `/path/to/table/_delta_log/{version:020d}`, and have that return all files whose embedded version is not smaller than `version`. So e.g. if `version=10` then the listing should return `/path/to/table/_delta_log/00000000000000000010.*` but also `00000000000000000011.*`, `00000000000000001000.*`, etc.

If setting `prefix=/path/to/table/_delta_log/` and `startFrom=/path/to/table/_delta_log/00000000000000000010` would achieve that, then we're in business. Except we'd need to arrange for `object_store` support of the same, which it perhaps lacks? The listing logic in Kernel's default engine just calls [list_with_offset](https://docs.rs/object_store/latest/object_store/trait.ObjectStore.html#method.list_with_offset), which states:
> Some stores, such as S3 and GCS, may be able to push offset down to reduce the number of network requests required.

So if listing is slow on Azure, most likely `object_store` doesn't fully take advantage of the Azure API?

**Comment by scovich** [comment]
@kyrre does the issue still repro for you? Or do recent versions of `object_store` perform well now?

**Comment by scovich** [comment]
> if listing is slow on Azure, most likely `object_store` doesn't fully take advantage of the Azure API?

I just became aware of https://github.com/apache/arrow-rs-object-store/pull/623
... which adds the missing support to Azure as of object_store 0.13.2
... which https://github.com/delta-io/delta-kernel-rs/pull/2116 brings into kernel as part of the arrow-58 upgrade
... which should be part of the next kernel release (0.21).

So this issue should self-resolve very soon.
